File: 30-days-of-code-challenges/day_18_queues_and_stacks.cpp
#include <iostream>
#include <queue>  
#include <stack>   

using namespace std;

class Solution {
    //Write your code here
    stack<char> s;
    queue<char> q;
    
    public: 
        void pushCharacter(char ch) {
            s.push(ch);
        }
        
        void enqueueCharacter(char ch) {
            q.push(ch);
        }
    
        char popCharacter() {
            return s.top();
        }
    
        char dequeueCharacter() {
            return q.front();
        }     
};

int main() {
    // read the string s.
    string s;
    getline(cin, s);
    
  	// create the Solution class object p.
    Solution obj;
    
    // push/enqueue all the characters of string s to stack.
    for (int i = 0; i < s.length(); i++) {
        obj.pushCharacter(s[i]);
        obj.enqueueCharacter(s[i]);
    }
    
    bool isPalindrome = true;
    
    // pop the top character from stack.
    // dequeue the first character from queue.
    // compare both the characters.
    for (int i = 0; i < s.length() / 2; i++) {
        if (obj.popCharacter() != obj.dequeueCharacter()) {
            isPalindrome = false;
            
            break;
        }
    }
    
    // finally print whether string s is palindrome or not.
    if (isPalindrome) {
        cout << "The word, " << s << ", is a palindrome.";
    } else {
        cout << "The word, " << s << ", is not a palindrome.";
    }
    
    return 0;
}

File: 30-days-of-code-challenges/day_18_queues_and_stacks.java
import java.io.*;
import java.util.*;

class Solution {
    // Write your code here.
    private LinkedList<Character> stack = new LinkedList<Character>();
    private LinkedList<Character> queue = new LinkedList<Character>();
    
    void pushCharacter(char ch) {
        stack.addLast(ch);
    }
    
    void enqueueCharacter(char ch) {
        queue.addLast(ch);
    }
    
    char popCharacter() {
        return stack.removeLast();
    }
    
    char dequeueCharacter() {
        return queue.remove(0);
    }
}

public class day_18_queues_and_stacks {
	
	public static void main(String[] args) {
        Scanner scan = new Scanner(System.in);
        String input = scan.nextLine();
        scan.close();

        // Convert input String to an array of characters:
        char[] s = input.toCharArray();

        // Create a Solution object:
        Solution p = new Solution();

        // Enqueue/Push all chars to their respective data structures:
        for (char c : s) {
            p.pushCharacter(c);
            p.enqueueCharacter(c);
        }

        // Pop/Dequeue the chars at the head of both data structures and compare them:
        boolean isPalindrome = true;
        for (int i = 0; i < s.length/2; i++) {
            if (p.popCharacter() != p.dequeueCharacter()) {
                isPalindrome = false;                
                break;
            }
        }

        //Finally, print whether string s is palindrome or not.
        System.out.println( "The word, " + input + ", is " 
                           + ( (!isPalindrome) ? "not a palindrome." : "a palindrome." ) );
    }
}